### ISSUE TITLE
Suggestion: Help, man pages, and/or info pages

### ISSUE THREAD

**Issue by Casteele72** [garglk/garglk#637]
There seems to be lack of dcumentation and help for actually running IF games. What command line arguments does it take? An keyboard shortcuts? Under X11, can I load a different game (there seems to be no menu bar or right click context menu)? I cannot figure out how to do anything such as save a transcript of my game, change games, execute special commands/actions (such as setting the initial random number seed, etc).

I cannot even figure out how to run Gargoyle in text only (TUI) mode, instead of X11 GUI.

**Comment by cspiegel** [comment]
These are some good points, and some proper documentation is definitely needed.  Let me at least address some of your concerns here:

> What command line arguments does it take?

None!  Well, since it's a Qt program it does accept standard Qt command-line args such as `-style`, but this is more incidental than anything else, and not particularly worth knowing.

> Any keyboard shortcuts?

Some.  Emacs-like cursor control is generally available (C-a to go to the beginning of the line, C-e to go to the end, C-d to delete a character, and so on).  C-c and C-v are copy and paste, as you'd expect.  C-, (control-comma) opens the config file in a text editor.

> Under X11, can I load a different game (there seems to be no menu bar or right click context menu)?

Not at the moment.  This is actually supported on macOS, but not on Windows/Unix.  Eventually I'd like to add such support to the Windows/Unix interface (they share the same one), and unify it with macOS. It's not a trivial task, though, since the launcher and the interpreters are completely separate. The launcher starts a new process to run the interpreters, and the interpreters have no concept of launching a game.  Eventually it'll probably happen, if only because I'd love to retire the Mac-specific interface code and replace it with Qt.  No time frame, though.

>  I cannot figure out how to do anything such as save a transcript of my game, change games, execute special commands/actions (such as setting the initial random number seed, etc).

"change games" is covered above, but the other options are, unfortunately, interpreter-specific.  Gargoyle makes use of third-party game interpreters, and itself only provides the user interface.  Which is to say, how all this happens differs between various interpreters and even games in the same interpreter.  With Z-machine games, for example, you can probably use the "SCRIPT" command to start a transcript, but nothing forces games to support it.  Setting a random seed is also a function of the interpreter, and may not even be supported, depending on the interpreter.

Now, all of this _can_ be documented.  But it's not quite as simple as it seems at first blush.

> I cannot even figure out how to run Gargoyle in text only (TUI) mode, instead of X11 GUI.

This is not possible.  Gargoyle is inherently a graphical program.  You can build all its interpreters against a terminal Glk implementation (see https://www.eblong.com/zarf/glk/; GlkTerm with widechar support is the one to choose).  But that'll take some effort. Some day I'd like to produce a Gargoyle-like project for the terminal, but it doesn't exist yet.

I will try to start jotting down stuff in the Wiki, to be consolidated, eventually, into a coherent document.  That is indeed sorely missing from Gargoyle.

If you're playing Z-machine games, there _is_ a manpage for Bocfel, the interpreter used for Z-machine games: https://cspiegel.github.io/bocfel/bocfel.html

**Comment by Casteele72** [comment]
If I understand correctly, [conceptually] Gargoyle itself is more of a front-end to other interpreters, by other people? (But packaged with/along-side Gargoyle?)

I have seen recent posts (here) regarding the flushing of the standard output stream. which is why I thought that Gargoyle itself could operate as both a TUI and a GUI application.

Perhaps, as a long-term goal, it could be made possible? My original purpose and  thinking was in making an archive (mainly for the original Infocom games) of game solutions, which could be launched something like 'gargoyle zork1.z3 --script my-solution.txt' or similar, and have it output the results to do anything I want with.

I would be happy to help when/where I can, both documentation and maybe some coding (although my coding skills are ancient, late 1990's style C/C++ -- but it seems to me many z-machine interps are from this era as well...). Let me know and maybe we can chat about it elsewhere without flooding the 'issues' section here.

**Comment by cspiegel** [comment]
> If I understand correctly, [conceptually] Gargoyle itself is more of a front-end to other interpreters, by other people? (But packaged with/along-side Gargoyle?)

Yes, that's completely correct.  To be more specific, Gargoyle includes its own [Glk](https://www.eblong.com/zarf/glk/) implementation, called garglk. Briefly, Glk is an API designed for interactive fiction. You can think of it as a portable I/O layer. Interpreters can target Glk instead of a specific OS, resulting in wide support across various platforms: there are Glk implementations for Unix, Windows, DOS, Android, etc, as well as specialized libraries like [RemGlk](https://www.eblong.com/zarf/glk/remglk/docs.html).

Gargoyle then packages a bunch of interpreters written by various authors, all of which use Glk, so get access to the Gargoyle UI.

> I have seen recent posts (here) regarding the flushing of the standard output stream. which is why I thought that Gargoyle itself could operate as both a TUI and a GUI application.

That was in relation to transcripts, which are implemented in terms of C's standard I/O library.

> Perhaps, as a long-term goal, it could be made possible? My original purpose and thinking was in making an archive (mainly for the original Infocom games) of game solutions, which could be launched something like 'gargoyle zork1.z3 --script my-solution.txt' or similar, and have it output the results to do anything I want with.

This would be best served by building interpreters against RemGlk or cheapglk.  There are some efforts to get Gargoyle's interpreters into a separate repository that's available to anybody who wants them, precisely so projects like you're describing can use them.  See https://github.com/cspiegel/terps and https://github.com/cspiegel/interpreters.  The latter is probably a better approach but not fully fleshed out yet.

As it stands, if you're looking to start work on a game solutions archive, I'd start by picking a single interpreter and using cheapglk (or RemGlk, if you need the extras it supports).  There's nothing in Gargoyle, apart from the interpreters themselves, which would be helpful, given that most of what Gargoyle provides itself is the Glk implementation.  If you're on a Unix-like system, the following will download and build Bocfel (the Z-machine interpreter Gargoyle uses) against cheapglk:

```sh
wget https://cspiegel.github.io/bocfel/downloads/bocfel-2.0.tar.gz
tar xfv bocfel-2.0.tar.gz
cd bocfel-2.0
wget https://www.eblong.com/zarf/glk/cheapglk-106.tar.gz
tar xfv cheapglk-106.tar.gz
(cd cheapglk && make -j`nproc`)
make GLK=cheapglk -j`nproc`
```
You now have a `bocfel` executable which uses stdin/stdout.  Use the `-r`, `-R`, and `-z` options to play back a command script while setting a PRNG seed:
```
./bocfel -z 1 -r -R sanddancer.txt sanddancer.z8
```
This will seed the PRNG with a value of 1, read a command script from `sanddancer.txt`, and run `sanddancer.z8`, dumping the output to stdout.  At this point it wouldn't be too difficult to write a small front-end to manage playback.

> I would be happy to help when/where I can, both documentation and maybe some coding (although my coding skills are ancient, late 1990's style C/C++ -- but it seems to me many z-machine interps are from this era as well...). Let me know and maybe we can chat about it elsewhere without flooding the 'issues' section here.

This is as good a place as any to discuss documentation improvements.  As an end-user who is asking for documentation, I'd say you're in a decent position to know what Gargoyle is missing.  What was the biggest "hole", so to speak, in understanding Gargoyle, from your perspective?  I have a hard time being able to determine levels of importance in missing documentation.  But if you can see an obvious deficiency and are willing to work on it, I'll answer any questions you have and help get things into shape.  You can add documents to the Wiki, and of course we can add things, such as man pages, to the repository.

I do think one area of Gargoyle that could use help is in its configuration.  Gargoyle is pretty configurable, but it's all done via a configuration file that's not necessarily very discoverable.  I'm not sure how many people are even aware there is such a thing.  And the config file itself is documented (via comments), but it may be useful to have a more comprehensive document explaining it.

Thanks for wanting to make Gargoyle better!

**Comment by Casteele72** [comment]
For the documentation... Start with what we have already brought up: What Gargoyle _is_, and just as importantly, what it is _not_. I installed it originally because it seemed to be the only multipurpose "interpreter" I could find for Linux -- Everything else I looked at seems to be aimed at other platforms (Mac, Win, Palm, etc.). But now that I understand that it is not an interpreter itself, but more of a front-end to others, I can better see how I can use it - and when I need to use something else, such as verifying and playback solutions.

I have installed fixmo, frtoz, and zoom from my distro's repo, but they are all becoming outdated as well for Debian. I will be manually building locally when I can, but for now, the pre-built ones seem to work for the solutions archive purpose (but lack graphics for playng some games). And I am still learning about interps other than for z-machine. bocfel does seem like a good one for z-machines as well; I just have not had enough time at one sitting to build it locally and it is not in my distro's repo.

For man pages, most of what has been said here should go in them. Command line options (which you said there are none, other than a game file to load), and let the user know that Gargoyle itself does not really provide an active UI once it passes the game off to the interp.

For the config file... Let's start with a simple question: Where is it and what is it named?

**Comment by cspiegel** [comment]
I'm going to address just one thing here initially:

> For the config file... Let's start with a simple question: Where is it and what is it named?

There are multiple answers to this.  First, it depends on the operating system.  And once you've identified the operating system, it can also depend on other factors (mainly historical).

The "best" way to determine the location is to start up a game and hit Control-comma (or Command-comma on Mac), to open up the config file in a text editor, then note where that is.  That's kind of gnarly, to be sure...

It'd probably be nice to have a command-line flag which provides this information, which would dovetail nicely with a man page, since now we'd actually _have_ at least one flag to document.

Gargoyle also allows games to override its settings (e.g. to set their own colors).  Really, what Gargoyle does is find a list of all possible config files, loading settings from all, resolving conflicts by giving each config file a priority.  Config files are also split into two categories, which are referred to as "user" and "non-user" config files.  User config files are those which the user can reasonably be considered the user's "main" config file, and one that he'd want to edit.  So it excludes game-specific config files as well as system-wide config files.

With that not-at-all-confusing introduction, here's how user config files are searched, from lowest to highest priority:

Windows:
* `<Current directory>\garglk.ini`
* `%APPDATA%\Gargoyle\garglk.ini`

Haiku:
* `<Settings dir>/Gargoyle`

All others (this is really Unix, including Mac):

* `$HOME/.garglkrc`
* `$HOME/.config/garglk.ini`
* `$XDG_CONFIG_HOME/garglk.ini`
* `$HOME/garglk.ini` (Mac only)

In truth, on Mac, `$HOME/garglk.ini` should probably be the lowest priority, since that pollutes $HOME.  But for now, here we are.  Again, historical baggage here, and not necessarily how it'd be done right now, starting from scratch.

**Comment by Aman9das** [comment]
Yeah this is a big issue, i had no idea gargoyle is themable before today. When I used play 6 years ago, i inverted the colours in my entire os to get dark mode. Lack of documentation is very painful.

**Comment by Casteele72** [comment]
This might be something I can help with, as well... even with coding.

My thinking is to simplify all this config madness, from the users perspective. Place one master config file in a known and predictable location. In turn, it can be used to generate per-interp config files "on the fly" as needed (and monitor them for changes to write back to the master config).

I would use XML, myself, since I have many tools and scripting languages which handle XML easily, and can pretty much import/export to anything else... Window's user used to *.ini format? No problem. Comma Separated Values (CSV)? No problem. Un*x style *.conf files? No problem. Original IBM-PCjr ROM BASIC byte code? Well... _that one_ might be a problem...

As for location... Typical software these days use some form of $HOME (likely with a additional directory appended, such as $HOME/.config on Linux).

I am even pondering how much it would take to embed something like Lua, Python, or TCL in Gargoyle, any of which can extend Gargoyle, and possibly any interps, in many ways with little additional code.

For example, embedding TCL brings with it the TK GUI graphics library -- Suddenly, Gargoyle and it's interps become full-fledged GUI apps, and can even play games with graphics and sound. TCL can also... Operate on Android devices (but not iOS :-/). Or offload the interps as TCL modules that can be loaded/unloaded dynamically (*.dll on Windows, *.so on Linux and Mac). How easy (or hard) any of this would be, I am not certain, as I am still learning how Gargoyle works, and have not looked at the code yet.

**Comment by Casteele72** [comment]
That was not quite what I meant. :-) I mean, as I see it, Gargoyle is more or less akin to a GUI version of a terminal emulator (such as Konsole, on my Debian Linux machine). But there are some drawbacks, including not being a full-fledged GUI but rather a console-type TUI presenting in a GUI window.

This presents a problem for authors, developers, interpreter implementations, et cetera: Presenting graphics and sound is not an (easily) viable option. (I am still learning about GLK, and how or even if GLK changes the scenario....) But to my thinking, what if Gargoyle can act as the "go-between" in these cases? Having a graphical output/window which can be drawn on, and handling dispatching audio to the sound system, etc.

Of course, any of this _is_ moot if such is completely beyond what you intended Gargoyle to be -- or too difficult to implement. I am just thinking as an end-user here; Testing solutions is just one of my purposes, but I also want to be able to play many games in the manner originally intended ("old shcool"). I can load many (if not all?) my old Infocom z-machine games in DosBox, but that presents other issues:
1) DosBox has it's own limitations;
2) It requires a properly configured DosBox; and
3) It requires executing the original game interpreter packaged in the game, which may be decades or more out of date.

In most cases, the data files can be "extracted" from these games and played in more "modern" interpreters with access to more modern system facilities, if such interpreters exist. None of the ones available for Debian Linux can use graphics or sound, which was why I got excited to see Gargoyle in my upstream repository and gave it a try.

I was previously using DosBox and things like the DOSZip interpreter or WinFrotz on my Windows virtual machine. But that requires constant switching between systems/emulators/virtual machines/interpreters/etc. I simply want a single, unified, consistent way to play different games, possibly translating certain functions as needed between the "presentation" (Gargoyle?) and the underlying interpreter, such as "<META>+S" invoking a "save state" function/feature.

I may not be wording all of these very clearly, of course. And as I learn about GLK, I wonder how much of this overlaps wih GLK itself. But it does seem to me that more often than not, the various components do not interoperate seemlessly from the end-user perspective. Maybe Gargoyle can lead the way towards making this happen and provide developer encouragement to provide more consistent interfacing/APIs in the interpeter back-ends...

I.E., let the interpreter developers worry more about the interpreter internals, and not so much about how to present it to the user. While Gargoyle and it's ilk can worry more about how to present it to the user without worrying about how the interpreter(s) work internally.

Of course, I may still be unclear myself on what Gargoyle _is_ and _is not_, or what it's intended niche/purpose is. Which goes back to my original post regarding documenting such things.

**Comment by Casteele72** [comment]
Unfortunately, the Infocom games you mentioned were the ones I was trying, so as you stated, I may be out of luck and need to do as you said (However, I cannot seem to get Winfrotz to work under Wine, only under a full Windows VM via VirtualBox). I have not tired many other IF games besides z-machine ones, yet, as my own focus right now is on Infocom game collections, but I will be expanding on others in the future.

You may also be correct that some of my idea/suggestions might be better sent to GLK development. I am still figuring a lo of things out -- Back to documentation, perhaps this is something that can be included. To explain... Are you familiar with the 7-layer OSI networking model? I am thinking in similar lines, structurally. Gargoyle sits at the top, same as FireFox, Chrome, or Edge, sit at the top of the TCP/IP stack. None of those worry, or care, about which NIC you physically have, which ISP you use, or which web  server you connect to (or even which OS you use, in many cases). GLK, as I see it right now, is akin to the TCP/IP stack _and_ the HTTP(S)/HTML protocols. And the various game engines/interpreters are akin to web servers. Games are simply like web pages; collections of data and resources (graphics and sounds). Each part only has to "worry" about how it is "connected" to the proper component(s), and conform to the protocol(s) in use. In this case, there is the added "complication" of there being a client-server model along side the rest.

From the end users perspective, they do not care about most of this. they just see it all as "the web". They want to view a web page, they go to the URL which provides it. They do not care if it is an Apache2 or NGINX web server providing the page. And so on.

Another example of my thinking is databases. Most DBMS systems (rBase, dBase, MS-JET, MySQL, MariaDB, etc.) provide db-system specific libraries to access and manipulate the database itself. But they also provide another method: SQL, a generalized and common language for database access, independent of both the database back-end and the application front-end that is in use by the user.

Those two examples provide a basic model of "how things work", and are common and generic enough that most users can understand them. Gargoyle's documentation can include such things. Give an overview of the entire system and then show where Gargoyle sits within the system. You will not need to provide much documentation about things like GLK or the interpreters other than to note which are in use/available (and let them provide their own more detailed documentation, as needed).

This would better help the user understand the system as a whole, and when to dig deeper in to Gargoyle documentation -- or to find and dig deeper in to other documentation (such as GLK) for what they want/how to do something.

Some other notes:

Regarding saves, I was specific in saying "save state" and not "save game" -- save the state of the interpreter (much like Frotz' "/save" command versus the game's own "save" ((without the "/" in front)) feature--if present) by serializing what is needed to save/restore it. This is most likely something I need to talk to GLK about, however -- Add specifications for in-band and out-of-band communication between the user and the game engines/interpreters. Something like, for example, GLK can get input from the user, "move north" (or just "n", etc.), and that can be passed to the game "in-band". While "out-of-band" can include things like a "user wishes to save and exit" event, which can signal the game (engine/interp) to generate and return a serialization, which, in turn, can be saved "locally". I will continue to learn about GLK, and likely submit such suggestions there where appropriate.

Regarding media (graphics and sound), that may also be a GLK thing. Set up/define a basic "virtual machine" style core to execute interpreters within. Various platforms (Apple II series, Commodore C64/128/Amiga, IBM PC-compatibles, etc.) often use specific ways to render the display (often memory mapped display frames) or produce sound (a bit more complex than display). The virtual machine can translate these between the "Guest" and "Host". Some interpreters already do this sort of thing internally, while others may require a lot of code refactoring. And some -- or much? -- of this is already covered by GLK/Glulx. (I still have not had much time to dig more in to that side.)

I had something else I waned to add here, but it is now after midnight here and my brain is sleepy and I work tomorrow. Good night.
